design-proposals/volume-update-strategy.md

```diff
@@ -0,0 +1,218 @@
+
+## Design details and motivation
+
+### Motivation on the approaches
+
+There a multiple ways to perform the storage migration using libvirt. It is possible to simply copy the storage to the new destination and then swap the original disk with the new one. This approach is the most efficient as the overhead is caused only by the storage copy.
+
+A different approach is to simultaneously move the storage and live-migrate the guest. When the destination storage is not available on the host where the VM is executing, this solution is required. For this case, costs are greater since there is storage as well as live VM migration overhead.
+
+As KubeVirt is designed today and due to Kubernetes design principles of volume pod immutability, it isn't possible to hotplug and unplug new volumes while the pod is still running. KubeVirt already supports volume hotplug by attaching a new volume to a dummy pod and then bind mounting the storage in the `virt-launcher` pod. Sadly, this method fails to address the issue of removing the old PVCs that were expressly declared in the virt-launcher pod.
+Declaring all volumes hotpluggable could be a potential remedy, but we prefer to avoid it because it has been shown to be fragile for some storage providers and opaque to Kubernetes.
+
+The only workable and generic solution for this restriction appears to be to live migrate the VM with storage. Live migration involves the creation of a new virt-launcher pod by KubeVirt that has the destination PVCs attached.
+
+This migrating to a new storage provider is an exceptional operation, and very resource intensive. If in the future, the k8s community will consider the volume mutability in the pod spec, then the lighter approach could be taken into consideration.
+
+## Limitations
+
+### Unsupported volumes
+
+Certain types of volumes and disks aren't suitable to be migrated or the feature isn't supported yet. Those migrated volumes will be marked as rejected in the VMI status
+
+The current types of rejected volumes are:
+  * Hotpluggable volumes, this case will be considered in a later iteration.
+  * Filesystem volumes, currently virtiofs doesn't support live-migration.
+  * Shareable volumes, if a disk is shared between multiple VMs, it cannot be safely copied.
+  * LUNs disks, originally the disk might support SCSI protocol but the destination PVC class does not. This case will be considered in a later iteration.
+  
+### Datavolumes
+
+DataVolumes are used to import the image during VM provisioning, but they may essentially be thought of as PVCs for the rest of the VM life cycle. With this API, we don't support live updates for the migration strategy to datavolumes. Therefore, the users can use this update strategy only when they replace datavolumes with PVCs.
+
+If the VM makes use of datavolume templates, then the template associated with the source volume needs to be removed by the user as well.
```

Comment:
Should we note that removing the template does not delete the datavolume?  Also if they want to delete the datavolume should wait until migration complete?

Reply:
I can add it